### PR TITLE
[SPARK-14541][SQL] Support IFNULL, NULLIF, NVL and NVL2

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?

This patch adds support for a few SQL functions to improve compatibility with other databases: IFNULL, NULLIF, NVL and NVL2. In order to do this, this patch introduced a RuntimeReplaceable expression trait that allows replacing an unevaluable expression in the optimizer before evaluation.

Note that the semantics are not completely identical to other databases in esoteric cases.
## How was this patch tested?

Added a new test suite SQLCompatibilityFunctionSuite.

Closes #12373.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 1]
```diff
@@ -222,6 +222,33 @@ trait Unevaluable extends Expression {
...
+  /**
+   * Method for concrete implementations to override that specifies how to construct the expression
+   * that should replace the current one.
+   */
+  def replaceForEvaluation(): Expression
+
+  /**
+   * Method for concrete implementations to override that specifies how to coerce the input types.
+   */
+  def replaceForTypeCoercion(): Expression
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: this is probably useful even outside RuntimeReplaceable, but I'm putting it only here for now. If we make this more generic, we can trim quite a bit of our type coercion code.

**Comment by rxin** [comment]
cc @hvanhovell wanna review this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 2]
```diff
@@ -222,6 +222,33 @@ trait Unevaluable extends Expression {
 
 
 /**
+ * An expression that gets replaced at runtime (currently by the optimizer) into a different
+ * expression for evaluation. This is mainly used to provide compatibility with other databases.
+ * For example, we use this to support "nvl" by replacing it with "coalesce".
+ */
+trait RuntimeReplaceable extends Unevaluable {
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: This is very cool!

**Comment by yhuai** [comment]
LGTM

**Comment by yhuai** [comment]
Merging to master and branch 2.0.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 2 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: Yea.

**Pull request closed without merging**